Source/JavaScriptCore/wasm/js/JSToWasm.cpp

```diff
@@ -381,6 +381,9 @@ std::unique_ptr<InternalFunction> createJSToWasmWrapper(CCallHelpers& jit, const
         unlinkedWasmToWasmCalls->append({ linkBuffer.locationOfNearCall<WasmEntryPtrTag>(call), functionIndexSpace });
     });
 
+    // Restore stack pointer after call
+    jit.addPtr(MacroAssembler::TrustedImm32(-static_cast<int32_t>(totalFrameSize)), MacroAssembler::framePointerRegister, MacroAssembler::stackPointerRegister);
```

Comment:
I would love to see tests stressing the case when we grow the stack due to a tail call to a leaf function. I feel like this might break our stack checking, since it is skipped for leaf functions? I'll have to dig in more

Reply:
I see. A leaf function will assume its caller did a check with extra size and will skip the check if its size is inferior to that size.
In B3 for example: https://github.com/WebKit/WebKit/blob/baab8dd057cfc07c8ba516883f0f1395bc0b1fda/Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp#L764

then:

https://github.com/WebKit/WebKit/blob/baab8dd057cfc07c8ba516883f0f1395bc0b1fda/Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp#L766

If we generate code for a function that tail calls, we can unconditionally do a stack check `(m_makesCall = true`), or we can try to optimize:
First we check if the function frame size is inferior to `minimumParentSize`, the we check if the difference between the previous frame and the shuffled frame **+ extraSize** is still inferior to `minimumParentSize`, otherwise we do a stack check.
Therefore, we both try to optimize and ensure our callee that we did a check. Does that make sense?
With this solution we would have to mind the situation in which JS calls Wasm tail calls Wasm: we would need to add a stack check in `jsToWasm` if the entrypoint ends up tail-calling, otherwise we cannot rely on `minimumParentSize` in the generators.

Reply:
This sounds right to me. There is no need to add a stack check in jsToWasm, because you can't call back into JS without a stack check. We have a soft stack limit for cases like this, and it is OK to go beyond it a bit. The soft stack limit is designed to make sure that c operation calls have enough stack space not to crash